UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs

```diff
@@ -589,6 +589,11 @@ protected override async Task OnOpenAsync(CancellationToken token = default(Canc
                 }
 
                 var userNameIdentity = (UserNameIdentity)this.UserIdentity;
+                if (userNameIdentity.UserName == null && userNameIdentity.Password == null)
+                {
+                    throw new Exception("Authentication is not anonymous, but no username and password were provided.");
```

Comment:
A plain `Exception` is very general. I think an `InvalidOperationException` could work here. According to its [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=netframework-4.8):

> The exception that is thrown when a method call is invalid for the object's current state.

Reply:
Changed to `InvalidOperationException`